### PR TITLE
Guard nil HeadObjectOutput fields

### PR DESCRIPTION
**Pull request by jcharum**
See https://github.com/grailbio/base/pull/14#issuecomment-622098407 for context.

### CODE REVIEW

**Review by josh-newman — APPROVED** [review]

File: file/s3file/s3file.go [hunk 1]
```diff
@@ -750,8 +750,16 @@ func (f *s3File) handleStat(req request) {
...
+		if output.ETag == nil || *output.ETag == "" {
+			req.ch <- response{err: errors.E("s3file.stat: empty ETag", f.name, errors.NotExist, "awsrequestID:", ids.String())}
+			return
+		}
+		if output.ContentLength == nil {
+			req.ch <- response{err: errors.E("s3file.stat: nil ContentLength", f.name, errors.NotExist, "awsrequestID:", ids.String())}
+			return
+		}
+		if output.LastModified == nil {
+			req.ch <- response{err: errors.E("s3file.stat: nil LastModified", f.name, errors.NotExist, "awsrequestID:", ids.String())}
```
Comment (tatatodd) [inline file/s3file/s3file.go]: This LGTM, thanks!

FYI there's similar logic in `s3Impl.Stat` around line 540 which should be updated in a similar fashion.

Comment (jcharum) [reply file/s3file/s3file.go]: Yep; you're right. I'll do that. Thanks again for flagging this!

**Comment by jcharum** [comment]
@josh-newman: Could you take another look? I applied the same guards to `(*s3Impl).Stat`, refactoring along the way.

File: file/s3file/s3file.go [hunk 2]
```diff
@@ -1396,6 +1351,49 @@ func ParseURL(url string) (scheme, bucket, key string, err error) {
...
+		if output.ETag == nil || *output.ETag == "" {
+			return nil, errors.E("s3file.stat: empty ETag", path, errors.NotExist, "awsrequestID:", ids.String())
+		}
+		if output.ContentLength == nil {
+			return nil, errors.E("s3file.stat: nil ContentLength", path, errors.NotExist, "awsrequestID:", ids.String())
+		}
+		if *output.ContentLength == 0 && strings.HasSuffix(path, "/") {
+			// Assume this is a directory marker:
+			// https://web.archive.org/web/20190424231712/https://docs.aws.amazon.com/AmazonS3/latest/user-guide/using-folders.html
+			return nil, errors.E("s3file.stat: directory marker at path", path, errors.Invalid, "awsrequestID:", ids.String())
```
Comment (josh-newman) [inline file/s3file/s3file.go]: It looks like this code used to be `errors.NotExist`?

Comment (jcharum) [reply file/s3file/s3file.go]: This is true, but it seemed like the wrong classification to me. I thought the risk of this actually breaking someone was very low, but it is backwards-incompatible.

Comment (josh-newman) [reply file/s3file/s3file.go]: I'd be concerned about breaking code that depends on this behavior for directory crawling, etc. It seems like the sort of thing that might not be well-integration-tested in some applications. Your call, though.

**Review by josh-newman — APPROVED** [review]

File: file/s3file/s3file.go (hunk 2 above)
Comment (jcharum) [reply file/s3file/s3file.go]: Okay; at a minimum it belongs in a separate PR.

**Comment by tatatodd** [comment]
Thanks very much folks for the fix!

If it's easy, can you create a `v0.0.8` release at some point?  We'll use a commit hash in our `go.mod` in the meantime, but it'd be nice to be on a tagged release.  :)

**Comment by jcharum** [comment]
> If it's easy, can you create a `v0.0.8` release at some point?

https://github.com/grailbio/base/releases/tag/v0.0.8

**Comment by tatatodd** [comment]
Awesome, thank you!

**Pull request merged**